### ISSUE TITLE
Performance degression between 1.1.1 and 1.1.2

### ISSUE THREAD

**Issue by x3cion** [wdavidw/node-csv-parse#103]
Hi there,

I'm using this module in a data import CSV -> MySQL for ~ 250.000 rows. Since 1.1.2 the performance degraded around 50 %. I cannot really pin it down, but maybe you have an idea. I cannot give out data, but here's the configuration for the parser:

`{
    "delimiter": "|",
    "rowDelimiter": "\n",
    "escape": "",
    "quote": "",
    "relax": true,
    "relax_column_count": true,
    "columns": true
}`

The profiler showed nothing suspicious either. If I can help in any way let me know!

**Comment by wdavidw** [comment]
must be related to #99. If a 50% performance penalty is confirmed we might roll back, @hakatashi what do u think?

**Comment by hakatashi** [comment]
I have no idea that confirms 50% performance degression. One possibility is that in the main loop I added the code block to check if the remaining buffer can be special sequence (https://github.com/wdavidw/node-csv-parse/pull/99/files#diff-f35c5932bb4da98c9da8192a6851d827R216). It includes some string subsettings and comparisons, but they are conditionally executed when the remaining buffer is short enough and should not cost too much...

**Comment by hakatashi** [comment]
Anyway we should analyze first before discussing when talking about performance.

**Comment by x3cion** [comment]
Small Test: https://gist.github.com/x3cion/c01b5f41dab09dd16466539d486e60c2

npm install csv-parse@1.1.2 to see the difference. 

Hope that helps!

**Comment by wdavidw** [comment]
out of curiosity, which profiling technique do u use?

**Comment by x3cion** [comment]
I used --prof on nodejs and analyzed it in PHPStorm, which separates the same function for being run optimized and unoptimized. While in 1.1.2 the function runs unoptimized all the time (1113x), it's often optimized (885x) in 1.1.1 or if changing this particular code block back. It still runs unoptimized 370x though, maybe that could be worked out too, since this function seems to be the heartpiece.

I also tried --trace-deopt. --trace-deopt shows "out of bounds" for the reason of deoptimization of Parser.__write. 

I guess the problem might not be the code block directly, but rather a misscalculation so the "look ahead space" the comment talks about is not big enough? I can't really tell, I didn't read the code, but it would be my first guess.
